[SYCL][Doc] Add specialization constant-length alloca extension proposal (#12660)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

private arrays whose size is given during the execution of the SYCL
application. There is no possible way of implementing this using `std::array`,
as the size of such container must be known at compile time, so we propose to
define a new `private_alloca` function whose size is specified using SYCL
specialization constants.

[NOTE]
====
This extension only supports SPIR-V backends for now, as it relies on
SPIR-V-specific capabilities,
such as specialization constants.
====

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
in the table below. Applications can test for the existence of this macro to
determine if the implementation supports this feature, or applications can test
the macro's value to determine which of the extension's features the
implementation supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

> === The `private_alloca` function

Reviewer: It would be nice if the definition of this function followed some of our more recent extensions, and defined its behavior using some of the [magic words](https://eel.is/c++draft/structure.specifications) from the ISO C++ specification.  That would help to address some ambiguities.

For example, as written, it's not clear if unsupported `ElementType` templates are hidden (i.e., the restriction is a _Constraint_) or if unsupported `ElementType` templates throw errors (i.e., the restriction is a _Mandate_).

Author: True. I think the text is cleaner now using the magic words. Thanks for the suggestion

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
template <typename ElementType, auto &SpecName,

>           access::decorated DecorateAddress>

Author: Until now, APIs in the spec use `legacy` as a default, something we may not want to do in a new extension. Also, there has not been a decision between `yes` and `no` on what the default should be in more recent APIs and conversations. Because of this, I'd rather keep it as is for now (although I agree having a default would be better) for future-proofing and only committing to a default when there is a consensus. WDYT?

>           access::decorated DecorateAddress>

Reviewer: I think we (as a group) need to come to an agreement.  Having to specify the element type, specialization constant name and decoration here all make this functionality less user-friendly.  @gmlueck -- is this something we should discuss at Khronos?

Since this is an experimental extension, I'd be okay with just adding an issue at the bottom saying that we'd like to default `DecorateAddress` to something, but we don't know what it should be.

>           access::decorated DecorateAddress>

Author: Added an issue. Thanks!

>           access::decorated DecorateAddress>

Reviewer 2: We should definitely discuss this default at Khronos, but following John's comment on the cast instruction https://github.com/intel/llvm/pull/12689#discussion_r1486672303 I guess this would also apply here (especially for the "harder to teach" part)

private_ptr<ElementType, DecorateAddress>

> private_alloca(kernel_handler &h);

Reviewer: I think it's problematic that this function is called `alloca` but doesn't really look or behave like `alloca`.  Compare:

```c++
void* bytes = alloca(N);
```
and
```c++
private_ptr<float> ptr = sycl::private_alloca<float, N>(kh);
```

It seems really unnatural to me that this function doesn't accept a size.  Could we use a design like this instead:
```c++
void* bytes = sycl::alloca(kh.get_specialization_constant_something<N>);
``` 

(In hindsight, having `get_specialization_constant` return a value directly may have been a mistake.  The intent of this `get_specialization_constant_something` placeholder is to return a representation of the specialization constant that's associated with a specific kernel handler, rather than the specialization constant's value.)

Reviewer: > For context: Our initial design had a different name for the function (allocate_private_memory). We decided to rename as we got comments suggesting people would understand the lifetime of the allocated memory better that way, as reading ...alloca would suggest that was stack allocated memory.

I agree that aligning with `alloca` will help people to understand what the function does.  But I think that alignment comes with a cost, and that we should strive to match the interface where possible.  Otherwise, we've increased the cognitive burden for programmers: people who already know how `alloca` works will read the code and probably think `kh` is a size, and learning the new behavior requires them to learn a new special-case.

> Returning a multi_ptr instead of void * was decided as not only it follows conventions set by other extensions, e.g., [sycl_ext_oneapi_local_memory](https://github.com/intel/llvm/blob/76a8d165a0e657231f3e80a227ad9ef672a650ce/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc), but also follows sycl::malloc<T>(size_t count, ...) and adds a layer of abstraction. Users will end up doing private_alloca(sizeof(ElementType) * ...), so we are saving that bit.

`sycl::malloc<T>` exists, but so does `sycl::malloc`.  Supporting both options might be worth considering.

> The decision on passing specialization_id as a template argument and kernel_handler & comes from the current specialization constant semantics in the spec. specialization_id, as you know, is just the specialization constant ID and there is no way of getting the specialization constant itself. What you propose, getting a specialization constant reference or similar, might be a better option, but I don't think we want to introduce such a big change to specialization constants in this small extension.

I have a slightly different perspective here, which is that extensions like these are the perfect place to think about what we could/should change in a future version of SYCL.  If we think that a new way of passing specialization constants to functions is desirable, we could prototype it here and move it to a separate extension later.

If you're worried about making changes to `kernel_handler` here, I realized you don't have to.  It might even be cleaner to define a new class, used like this:
```c++
auto size = sycl::spec_constant<Size>(kh);
auto ptr = sycl::alloca(size);
```
...which should allow the element type and specialization constant templates to be deduced.  I think something like this would be much cleaner than a load of required template arguments, especially in the aligned case.

Again, I don't mean to block the initial implementation and experimentation work here, so adding an issue about this would satisfy me.

Tagging @gmlueck for awareness, since I know he has done a lot of work on specialization constants.

Author: I added an issue for the `specialization_constant` interface. I'll address the `void *` returning interface and aligned version of the function tomorrow

Reviewer: My thinking is that it would be nice not to tie this function to `multi_ptr`, because lots of codes use only raw pointers.

Am I right in thinking that `private_ptr` can be implicitly cast to the underlying pointer type?  That would be enough to enable the usage pattern I have in mind.

Reviewer 2: > Am I right in thinking that private_ptr can be implicitly cast to the underlying pointer type?

you are (like any `multi_ptr`) but it is deprecated in 2020

Author: `void *ptr = private_alloca<bytes>(kh).get_raw()` (regardless of default) isn't that bad

Author: @Pennycook We have been giving this a thought and maybe returning a `void` pointer to align with C-like interfaces is not ideal. With the current proposal (missing an additional signature with alignment), users can pass `std::byte` as an `ElementType` and do type punning as they wish in case that's needed. Do you think we could leave this as is and only add a `void` pointer signature if users come up with an actual use-case for this?

This is an experimental feature anyway, so we could extend it in the future in cases like that.

Reviewer: Yes, that's fine by me.

} // namespace sycl::ext::oneapi::experimental
----

This extension adds a new `private_alloca` function that can be used to allocate
a private memory region with capacity for
`h.get_specialization_constant<SpecName>()` elements of type `ElementType`. The
returned pointer will be aligned for `ElementType`.

The underlying memory region is automatically freed when the caller to
`private_alloca` is returned.

In case of private memory exhaustion, the underlying backend must report an
error in the same fashion as if the allocation size were static.

> If this function is called from host context or an unsupported backend, an

Reviewer: Just checking -- this means that the function needs to be defined for all backends, even if it doesn't work.  Could/should this functionality be hidden behind an aspect instead?

Author: I defined an aspect for that. Thanks!

`exception` with the `errc::feature_not_supported` error code must be thrown.

> `ElementType` must be a cv-unqualified trivial type. The return memory is

Reviewer: ```suggestion
`ElementType` must be a cv-unqualified trivial type. The returned memory is
```

  memory region has capacity for `h.get_specialization_constant<SpecName>`
  `ElementType` elements. The default value for the specialization constant must
  be at least one and the specialization constant must not be set to a value
  less than one. Setting the specialization constant to a value less than 1 or
  providing a default value less than 1 is undefined behaviour.
`DecorateAddress`:: Whether the returned `sycl::multi_ptr` is decorated or not.

==== Return Value

`sycl::private_ptr` to a region of `h.get_specialization_constant<SpecName>()`
elements of type `ElementType` aligned for such type. The underlying memory
region will be automatically deallocated when the function from which
`private_alloca` is called returns.

== Example usage

This non-normative section shows some example usages of the extension.

[source,c++]
----
constexpr specialization_id<int> size(1);

class Kernel;

SYCL_EXTERNAL void impl(const float *in, float *out, size_t n,
                        decorated_private_ptr<float> ptr);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);

>     h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler h) {
>       // Allocate memory for 'n' 'float's
>       auto ptr = private_alloca<float, size, access::decorated::yes>(h);
>       // Use pointer in implementation
>       impl(in, out, h.get_specialization_constant<size>(), ptr);

Reviewer: Reusing the name `h` is legal, but confusing in this case.  I suggest renaming the kernel handler:

```suggestion
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      // Allocate memory for 'n' 'float's
      auto ptr = private_alloca<float, size, access::decorated::yes>(kh);
      // Use pointer in implementation
      impl(in, out, kh.get_specialization_constant<size>(), ptr);
```

};
} // namespace sycl
----

The `ext_oneapi_scla` aspect indicates that the device is capable of using the
`private_alloca` API defined in the following sections.

=== The SCLA API

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
template <typename ElementType, auto &SpecName,
          access::decorated DecorateAddress>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);

template <typename ElementType, std::size_t Alignment, auto &SpecName,
          access::decorated DecorateAddress>
private_ptr<ElementType, DecorateAddress>
aligned_private_alloca(kernel_handler &kh);
} // namespace sycl::ext::oneapi::experimental
----

_Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`
must be a reference to a specialization constant of integral `value_type`. In
the case of `aligned_private_alloca`, `Alignment` must be a power of 2
fundamental alignment stricter than the alignment requirement of `ElementType`.

_Effects_: `h.get_specialization_constant<size>()` elements of type
`ElementType` are allocated and default initialized in private memory.

_Returns_: A pointer to a default initialized region of private memory of

> `h.get_specialization_constant<size>()` elements of type

Reviewer 3: Checking that I understand this correctly ... you're saying here that the returned pointer always points to a buffer if "size" elements, where "size" is the value of the specialization constant.  Correct?  That must mean that every call to `private_alloca` return the same pointer value.  For example, every iteration of the loop below returns the same value for `ptr`:

```
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      for (int i = 0;  i < N;  i++) {
        auto ptr = private_alloca<float, size, access::decorated::yes>(kh);
        // All iterations get the same value for "ptr"
    });
  });
}
```

Do I understand correctly?

If so, I think the name "alloca" is confusing because it implies that each call returns a separate allocation.  If my understanding is correct, it seems like this functions is pretty similar to [sycl_ext_oneapi_work_group_specific](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_specific.asciidoc), except that it returns memory in the private address space rather than in the local address space.

I have some more comments to make, but I'll stop here to check that my understanding is correct.

Author: You're right to some extent. `private_alloca` would be something in between, as we can have:
```c++
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      auto ptr0 = private_alloca<float, size, access::decorated::yes>(kh);
      auto ptr1 = private_alloca<float, size, access::decorated::yes>(kh);
      // ptr0 and ptr1 point to different allocations
    });
  });
}
```

`private_alloca` effectively declares an array. At runtime, after specialization constants are instantiated, your code would be equivalent to
```c++
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      for (int i = 0;  i < N;  i++) {
        float ptr[/*constant n value*/];
        auto ptr = private_alloca<float, size, access::decorated::yes>(kh);
        // All iterations get the same value for "ptr"
    });
  });
}
```

Maybe `alloca` is not the best word to describe these semantics...

Author: Do you think something like:

```c++
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      for (int i = 0;  i < N;  i++) {
        auto ptr = private_array<float, size, access::decorated::yes>(kh);
        // All iterations get the same value for "ptr"
    });
  });
}
```
i.e., renaming to `private_array` would make this somewhat clearer?

Reviewer 3: Maybe I'm missing something, but it seems like it would be much clearer to expose this as a class, rather than a free function.  For example:

```
constexpr specialization_id<int> size(1);

void run(queue q, const float *in, float *out, size_t n) {
  q.submit([&](handler &h) {
    h.set_specialization_constant<size>(n);
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      work_item_array<float, size> array_a{kh};
      work_item_array<float, size> array_b{kh};

      // array_a and array_b have different storage
      array_a[0] = 3.14f;
      array_b[0] = 2.718f;
    });
  });
}
```

This seems a lot clearer because each variable of type `work_item_array` has it's own unique storage.  With the free function, there are confusing rules about when the storage is unique.  Two dynamic calls to the same static instance of the function return the same storage, while calls to different static instances of the function return unique storage.

I also see it as an advantage that the syntax above is similar to the [sycl_ext_oneapi_work_group_specific](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_specific.asciidoc) extension.  It's easier for people to learn SYCL if we use the same pattern in multiple places, rather then using different patterns everywhere.

Is there some problem I'm missing with the `work_item_array` pattern I show above?

Author: > Maybe I'm missing something, but it seems like it would be much clearer to expose this as a class,

We also looked into options to have that as a class (including a similar syntax to `sycl_ext_oneapi_work_group_specific`. From the proposal:

> The big design constraint stems from the unknown allocation size at compile
time. C++ does not support variable length arrays and complete type sizes must
be known at compile time. Thus, the free function interface returning a pointer
to private memory is the better way to represent this construct in C++. Lifetime
of the underlying memory region was a concern too, but the current design with
automatic storage duration for the allocated memory region closely follows what
the user would get from a stack-allocated array.

SCLA as a dynamically sized type was ruled out then. A more similar design to the extension you mention was discarded also, due to restrictions in private memory allocations in underlying APIs. Effectively, the biggest difference between SYCL private memory and other kind of memories is that it is allocated in the call stack conceptually in underlying backends, e.g., in SPIR-V, it corresponds to [_Function Storage Class_](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_storage_class):

> Visible only within the declaring function of the current invocation.

Users writing something like:

```c++
work_item_specific<int[]> foo() {
  return {};
}
```

Where the returned pointer is deallocated as soon as `foo` returns, would run into some surprises seeing RAII violated.

Summing up: Private memory specifics made dynamic memory allocations really tricky in SYCL. We chose not to have this as a type and either violate C++ type rules and have a dynamically sized type or violate RAII. The alternative we found was the free function design, documenting that the pointer points to an automatic storage array.

I hope this throws some light in the design process we went through and why we ended up with this (imperfect) design.

Reviewer 3: We met offline.  Marking this conversation as resolved.

application, but that will not change when a SYCL kernel function is
invoked. This way, specialization constants could be used to implement SYCL
private arrays whose size is given during the execution of the SYCL
application. There is no possible way of implementing this using `std::array`,
as the size of such containers must be known at compile time, so we propose to
define a new `private_alloca` function whose size is specified using SYCL
specialization constants.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
in the table below. Applications can test for the existence of this macro to
determine if the implementation supports this feature, or applications can test
the macro's value to determine which of the extension's features the
implementation supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

> === New Aspect for Specification Constant-Length Allocations

Reviewer 3: ```suggestion
=== New aspect for specialization constant-length allocations
```

Typo on "specialization".  Kind of a nit, but section titles are normally lower case except for the first word.

This extension adds a new device aspect:

[source,c++]
----
namespace sycl {
enum class aspect : /*unspecified*/ {

>   ext_oneapi_scla

Reviewer 3: We generally try to avoid acronyms for SYCL identifiers, preferring to spell out words when possible.  The name "specialization constant length array" is pretty long, and it doesn't match the name of the API that allocates these things.  Could we use the name "private alloca" as the name of the extension?  This would mean:

* Document name: "sycl_ext_oneapi_private_alloca.asciidoc"
* Feature test macro: `SYCL_EXT_ONEAPI_PRIVATE_ALLOCA`
* Aspect: `ext_oneapi_private_alloca`

Author: Renaming applied :+1: More consistent now

};
} // namespace sycl
----

The `ext_oneapi_scla` aspect indicates that the device is capable of using the
`private_alloca` API defined in the following sections.

=== The SCLA API

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
template <typename ElementType, auto &SpecName,
          access::decorated DecorateAddress>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);

template <typename ElementType, std::size_t Alignment, auto &SpecName,
          access::decorated DecorateAddress>
private_ptr<ElementType, DecorateAddress>
aligned_private_alloca(kernel_handler &kh);
} // namespace sycl::ext::oneapi::experimental
----

> _Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`

Author: A [trivial class](https://timsong-cpp.github.io/cppwp/class.prop#2) must also be [trivially copyable](https://timsong-cpp.github.io/cppwp/class.prop#1), and that implies having a [trivial destructor](https://timsong-cpp.github.io/cppwp/class.prop#1.3). "Trivial type" is the correct term to use here IMO.

> _Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`

Reviewer 3: Agreed.

must be a reference to a specialization constant of integral `value_type`. In
the case of `aligned_private_alloca`, `Alignment` must be a power of 2

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Reviewer 3: I still can't parse this sentence.  The confusing part is "power of 2 fundamental alignment".  This makes it sound like the value must be the type's fundamental alignment raised to some power of 2.  Or, that the value must be 2^x where "x" is the type's fundamental alignment.  Neither of those things make any sense, though.

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Author: I dropped the power of 2 claim. Alignments must be powers of 2 by definition, so the current statement works. If you like it, we can keep the current one or use the alternative equivalent:

> ... `Alignment` must be a fundamental alignment stricter than the alignment requirement of `ElementType`.

Both work for me

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Author: I changed the condition to allow more values here. We no longer require fundamental alignments.

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Reviewer 4: > @v-klochkov does the current state look good to you?

The current design looks Ok/good to me.

Just some addition/idea. Originally I thought about something like this:
```C++
// Assuming 'DecoratedAddress' parameter may have default value.
// If DecorateAddress must be specified always, then the code is same, just remove '= yes' below.
template <typename ElementType, auto &SpecName,
          access::decorated DecorateAddress = yes,
	  std::size_t Alignment = std::alignment_of_v<ElementType>>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);

template <typename ElementType, auto &SpecName,
          std::size_t Alignment = std::alignment_of_v<ElementType>,
	  access::decorated DecorateAddress = yes>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);
```
The reasoning:
`ElementType` and `SpecName`(aka Size) must always be specified and cannot be skipped.
The API variant shown above would be convenient if someone already has a call
```C++
auto P1 = private_alloca<float, size>(h);
auto P2 = private_alloca<float, size, access::decorated::no>(h);
```
and need to add alignment, then it is enough to simply it as 3rd or 4th template param without renaming:
```C++
// Now all pointers are 128-byte aligned.
auto P1 = private_alloca<float, size, 128>(h);
auto P2 = private_alloca<float, size, access::decorated::no, 128>(h);
auto P3 = private_alloca<float, size, 128, access::decorated::no>(h);
```

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Reviewer 4: Having a different name `aligned_private_alloca` has the advantage of making the aligned-ness extra highlighted. So, either way is good to me.

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Author: I went with the `aligned_*` signature to follow SYCL and C++ existing library functions. I don't have a strong preference here either. @sommerlukas @Naghasan any preference?

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Reviewer 5: ```c++
// Assuming 'DecoratedAddress' parameter may have default value.
// If DecorateAddress must be specified always, then the code is same, just remove '= yes' below.
template <typename ElementType, auto &SpecName,
          access::decorated DecorateAddress = yes,
	  std::size_t Alignment = std::alignment_of_v<ElementType>>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);

template <typename ElementType, auto &SpecName,
          std::size_t Alignment = std::alignment_of_v<ElementType>,
	  access::decorated DecorateAddress = yes>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);
```

If I understand you @v-klochkov  correctly that you would propose to have both templates, this unfortunately wouldn't work due to ambiguous call in the case of `auto P1 = private_alloca<float, size>(h);` (https://godbolt.org/z/93dn19Gfj). 

I think we should go with the `aligned_*` signature as it follows C++ & SYCL and also makes the aligned-ness explicit.

> fundamental alignment stricter than the alignment requirement of `ElementType`.

Author: Okey, looking at your proposal a second time: having two signatures with the same name and your proposed usages would not work, as that would lead to ambiguous calls. As per the reasons I [exposed above](https://github.com/intel/llvm/pull/12660#issuecomment-1942112557), the two signatures approach might be more future proof. WDYT?

_Effects_: `h.get_specialization_constant<size>()` elements of type
`ElementType` are allocated and default initialized in private memory.

_Returns_: A pointer to a default initialized region of private memory of
`h.get_specialization_constant<size>()` elements of type
`ElementType`. `DecorateAddress` defines whether the returned `multi_ptr` is
decorated. In the case of `private_alloca`, the pointer is suitably aligned for
an object of type `ElementType`. In the case of `aligned_private_alloca`, the
pointer is aligned to the specified `Alignment`.

> _Remarks_: In case of private memory exhaustion, the underlying backend must

Reviewer 3: ```suggestion
_Remarks_: In case of private memory exhaustion, the implementation must
```

Author: "implementation" could be misinterpreted here, as that could mean the "runtime implementation". Maybe saying "the backend" instead is better?

Reviewer 3: I don't like using the term "backend" here for two reasons.  First, it seems to imply a specific implementation detail -- that memory exhaustion must be handled in the backend rather than some other part of the implementation.  Second, I can imagine SYCL implementations that do not use any backend.  For example, it's not clear what the term "backend" means for a purely native implementation of SYCL.

I see that you've already reworded this to use the term "implementation".  Let's keep that.  We use the same term "implementation" in other parts of the SYCL spec.

report an error in the same fashion as if the allocation size were static. In
case of a successful call, allocated memory has automatic storage
duration. Additionally, `SpecName` must have a default value of at least 1 and
not be set to a value less than 1 during program execution. Violation of these

> conditions results in undefined behaviour.

Reviewer 3: This last sentence about `SpecName` seems like a precondition.  Could we move it to a _Precondition_ section instead?  There is no need to say that violation of the precondition results in UB because that is the definition of a precondition.

Reviewer 3: Makes sense.
